### ISSUE TITLE
Access wrapped component

### ISSUE THREAD

**Issue by EdonGashi** [faceyspacey/react-universal-component#119]
Hi, is it possible to access wrapped component such as:

```js
const Universal = universal(() => import('MyComponent'))

// after some time, I need to do some actions based on static properties of the original component
if (Universal.wrappedComponent && Universal.wrappedComponent.doSomething) {
  // do something
}

// I cannot rely on hoisting, because the imported component may not have that static declared
if (Universal.doSomething) // I can't be sure if the static is missing or if the component has not yet loaded
```

I know about the callbacks, I was just wondering if there is a shorter way for this... Thanks!

**Comment by matthew-dean** [comment]
I have the similar question, but related to instance methods, not static methods. I'm wrapping `react-player` like this:

```js
import universal from 'react-universal-component'
const ReactPlayer = universal(() => import(/* webpackChunkName: 'library' */ 'react-player'), {
  resolve: () => require.resolveWeak('react-player'),
  chunkName: 'library',
  loading: emptyNode
})
```
And I'm trying to set a reference to the player, so I can call methods, like the example here: https://github.com/CookPete/react-player/blob/master/src/demo/App.js#L101-L103

The problem is that `ref` attaches a reference to the ProxyComponent rather than passing / attaching a reference to `react-player`. The ProxyComponent doesn't seem to maintain any instance reference to the wrapped component, making the instance methods inaccessible.

How do we use the `ref` React property with this library?

**Comment by EdonGashi** [comment]
There is some new API with refs, haven't look at it in depth yet, but maybe it can help you with your problem: https://reactjs.org/docs/forwarding-refs.html

**Comment by EdonGashi** [comment]
Another annoyance is that hoisting happens only when a render has been initiated, I cannot access static properties before the initial render.

Any chance for a preloadWeak() that returns something synchronously or hoists the component if it's loaded?

**Comment by matthew-dean** [comment]
@EdonGashi That helps, but `react-universal-component` itself would need to (and should) implement the `forwardRef` API to pass along refs to components it wraps. There's never any need to access the universal component itself via a `ref` assignment.

**Comment by zackljackson** [comment]
Yeah it is. I literally coded the feature for this exact thing. Update the Babel loader

**Comment by EdonGashi** [comment]
Yes, this is not a bug report, more like a feature request.

Universal component does not check if the wrapped component is available sync before starting an initial render cycle. That is okay for instance side, but for static side you won't be able to access statics without starting to render and hook onAfter.

I wanted to implement something like next's getInitialProps where I get some data which will be passed as props to the component itself. If I have to mount it first then I am forced to do a double render from server side, because some static methods may have side effects or may result in route changes (new render).

To have full control I wanted to do some static analysis on the component and then render it. Doing side effects during mounting does not feel right, especially in the server where you are in no hurry to render anything except the final view.

I will see about testing the PR'd code. Matthew also requested a feature, so we should see about fixing child refs too.

**Comment by zackljackson** [comment]
I’m currently working on some much desired Webpack 4 upgrades for the universal family. After that’s done, I can try and look into this, if you have the drive to lay the groundwork on the refs related stuff- we can try to collaborate on it. 

There’s a lot happening with Some big upgrades that we are gonna release, time constraints are hard to manage 😋

**Comment by EdonGashi** [comment]
Sure, I'll do whatever I can to help. Good luck on the webpack 4 stuff!

**Comment by zackkirsh** [comment]
Thanks! Last thing to knock out the park: https://github.com/faceyspacey/extract-css-chunks-webpack-plugin/issues/60

**Comment by zackljackson** [comment]
If you can make a PR and a test, I’ll merge it for you. You might get it done quicker than I can right now

**Comment by matthew-dean** [comment]
@zackljackson Fair enough! I'll see how critical this is to our project. If it is, there might be time for it.

**Comment by zackljackson** [comment]
@matthew-dean lets open a separate issue for ref forwarding, ive seen it around so it might be simple to do.

**Comment by matthew-dean** [comment]
@zackljackson Okay, do you need me to do that or were you going to do that?

**Comment by OliverJAsh** [comment]
Did anyone open an issue to track ref forwarding?

**Comment by ScriptedAlchemy** [comment]
dont think so
